File: 2023/day25/solution.py
#!/usr/bin/env python3
import sys
from collections import defaultdict
from itertools import combinations
from random import choice as randomchoice
from typing import TypedDict

# yapf: disable
sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i': FILENAME = 'input.txt'
      case '-p1': PART1, PART2 = True, False
      case '-p2': PART1, PART2 = False, True
      case '-p0': PART1, PART2 = False, False
      case _: raise Exception(f'unknown {arg=}')

def debug(level: int, *args, **kwargs):
  if level < DEBUG: print(*args, **kwargs)
#yapf: enable

with open(FILENAME, 'r') as f:
  data = {parts[0]: {*parts[1].split()} for line in f.read().splitlines() if (parts := line.split(': '))}


def karger_min_cut(edges: list[tuple[str, str]]):

  class ContractedElem(TypedDict):
    parent: str
    rank: int

  contracted: dict[str, ContractedElem] = {y: {'parent': y, 'rank': 0} for x in edges for y in x}

  def get_parent_id(vertex_id: str) -> str:
    # recursively find the parent of vertex_id
    if contracted[vertex_id]['parent'] != vertex_id:
      contracted[vertex_id]['parent'] = get_parent_id(contracted[vertex_id]['parent'])
    return contracted[vertex_id]['parent']

  remain_count = len(contracted) - 2
  while remain_count:
    # non-deterministic ugh
    edge = randomchoice(edges)

    left = get_parent_id(edge[0])
    right = get_parent_id(edge[1])
    if left == right: continue

    # parent one to the other based on rank
    rank_diff = contracted[left]['rank'] - contracted[right]['rank']
    if rank_diff < 0:
      contracted[left]['parent'] = right
    elif rank_diff > 0:
      contracted[right]['parent'] = left
    else:
      contracted[right]['parent'] = left
      contracted[left]['rank'] += 1

    remain_count -= 1

  cuts = [x for x in edges if get_parent_id(x[0]) != get_parent_id(x[1])]
  debug(1, f'{cuts=}')
  return cuts


def part1():
  # i looked up kargers min cut algo
  edges = []
  for left, rights in data.items():
    for right in rights:
      edges.append((left, right))
      # edge A,B == edge B,A
      # # edges.append((right,left))
  cuts: list[tuple[str, str]] = []
  # cuts=[('xzz', 'kgl'), ('xxq', 'hqq'), ('vkd', 'qfb')]
  while len(cuts) != 3:  # annoying
    cuts = karger_min_cut(edges)
  debug(0, f'3 cuts: {cuts}')

  # edges is not the best format for mapping so refactor
  connections: dict[str, set[str]] = defaultdict(set)
  for k, v in data.items():
    connections[k].update(v)
    for n in v:
      connections[n].add(k)
  # remove the cuts
  for cut in cuts:
    connections[cut[0]].remove(cut[1])
    connections[cut[1]].remove(cut[0])

  def map_network(start_node: str):
    found_nodes: set[str] = set()

    def map_nodes(from_node: str):
      if from_node in found_nodes: return
      found_nodes.add(from_node)
      for to_node in connections[from_node]:
        map_nodes(to_node)

    map_nodes(start_node)
    debug(0, f'{start_node=} {len(found_nodes)=}')
    return found_nodes

  all_nodes = {x for x in data.keys()}
  all_nodes.update(y for x in data.values() for y in x)
  debug(0, f'{len(all_nodes)=}')

  network_a = map_network(list(all_nodes)[0])
  network_b = map_network(list(all_nodes.difference(network_a))[0])
  print(f'part 1: {len(network_a)*len(network_b)}')

  # 514794

if PART1: part1()

File: 2023/day25/solution.ts
#!/usr/bin/env -S deno --allow-read
// #region base aoc template
const Maths = Math;

class Args {
  constructor(public filename = 'input.txt', public debug = 0, public part1 = true, public part2 = true) {
    for (const arg of Deno.args) {
      const [key, value] = [arg.slice(0, 2), (fallback = 0) => Number(arg.slice(2) || fallback)];
      if (key === '-d') debug = value(1);
      else if (key === '-e') filename = `example${arg.slice(2)}.txt`;
      else if (key === '-i') filename = 'input.txt';
      else if (key === '-p') [part1, part2] = [(value(0) & 1) === 1, (value(0) & 2) === 2];
      else throw new Error(`unrecognised arg="${arg}"`);
    }
    [this.filename, this.debug, this.part1, this.part2] = [filename, debug, part1, part2];
    console.log(`args: {filename: "${filename}", debug: ${debug}, part1: ${part1}, part2: ${part2} }`);
  }
}

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (args.debug >= level) console.debug(...data);
};

const args = new Args();
// #endregion

interface Pair {
  left: string;
  right: string;
}

const parseInput = () =>
  // parse as a sorted array of sorted pairs
  Deno.readTextFileSync(args.filename)
    .split('\n')
    .filter((line) => line.trim())
    .map((line) => line.split(': '))
    .flatMap(([left, rights]) => rights.split(/\s+/).map<Pair>((right) => ({ left, right })))
    .map(({ left, right }) => (left.localeCompare(right) <= 0 ? { left, right } : { left: right, right: left }))
    .toSorted((a, b) => a.left.localeCompare(b.left) || a.right.localeCompare(b.right));

const kargerMinCut = (connections: Array<Pair>) => {
  // python implementation was loosely based on: https://www.geeksforgeeks.org/introduction-and-implementation-of-kargers-algorithm-for-minimum-cut/
  const collapsedMap = new Map<string, { parent: string; rank: number }>(
    connections.flatMap((item) => [item.left, item.right]).map((item) => [item, { parent: item, rank: 0 }])
  );
  const getParent = (node: string): string =>
    collapsedMap.get(node)!.parent === node ? node : getParent(collapsedMap.get(node)!.parent);
  let remainCount = [...collapsedMap.keys()].length - 2;
  while (remainCount > 0) {
    const pair = connections.at(Maths.floor(Maths.random() * connections.length))!;
    const [left, right] = [getParent(pair.left), getParent(pair.right)];
    if (left === right) continue;
    const rankDiff = collapsedMap.get(left)!.rank - collapsedMap.get(right)!.rank;
    if (rankDiff < 0) collapsedMap.get(left)!.parent = right;
    else if (rankDiff > 0) collapsedMap.get(right)!.parent = left;
    else {
      collapsedMap.get(right)!.parent = left;
      ++collapsedMap.get(left)!.rank;
    }
    --remainCount;
  }
  return connections.filter((item) => getParent(item.left) !== getParent(item.right));
};

const mapNetwork = (connectionMap: Map<string, Array<string>>, startNode: string) => {
  const network = new Set<string>([startNode]);
  const walk = (node: string) =>
    // for/of is faster but this isn't performance critical
    connectionMap
      .get(node)!
      .filter((neighbour) => !network.has(neighbour))
      .forEach((neighbour) => {
        network.add(neighbour);
        walk(neighbour);
      });
  walk(startNode);
  return network;
};

const part1 = () => {
  const connectionPairs = parseInput();
  debug(3, connectionPairs);
  // karger min cut is non-deterministic so we just have to brute force it until we get the required 3 cuts :|
  let cuts: Array<Pair>;
  while ((cuts = kargerMinCut(connectionPairs)).length !== 3) debug(2, { cuts });
  debug(1, 'found', { cuts });

  // filter cuts from the pairs and convert into a map of arrays
  // the filtering would be faster if connectionPairs was a set, but the brute force above is the slow part
  const connectionMap = new Map<string, Array<string>>();
  connectionPairs
    .filter((item) => !cuts.some((cut) => cut.left === item.left && cut.right === item.right))
    .forEach((item) => {
      connectionMap.has(item.left) ? connectionMap.get(item.left)!.push(item.right) : connectionMap.set(item.left, [item.right]);
      connectionMap.has(item.right) ? connectionMap.get(item.right)!.push(item.left) : connectionMap.set(item.right, [item.left]);
    });
  debug(3, { connectionMap });
  const nodes = new Set<string>(connectionMap.keys());

  // walk the network, return reachable nodes
  const leftNetwork = mapNetwork(connectionMap, [...nodes.values()][0]);
  debug(3, { leftNetwork });
  const rightNetwork = mapNetwork(connectionMap, [...nodes.difference(leftNetwork)][0]);
  debug(3, { rightNetwork });

  const result = leftNetwork.size * rightNetwork.size;
  console.log('part 1:', result);
};

const part2 = () => {};

if (args.part1) part1();
if (args.part2) part2();
